File: web/protocol.js
// Binary datagram format shared with server/protocol.py (big-endian).
// DATA: type u8 | flow u8 | seq u32 | send_ts f64 (ms) | padding
// ACK:  type u8 | flow u8 | seq u32 | echo_send_ts f64 | recv_ts f64
// ACK_BLOCK: type u8 | flow u8 | ack_id u32 | largest u32 | largest_recv_ts f64 | ack_delay_ms f32 |
//            low u32 | n_ranges u8 | (first u32, last u32)* | n_ts u16 | (seq u32, recv_offset_ms f32)*
//   See server/protocol.py for semantics.
// `flow` is the direction of the DATA a packet refers to.

export const DATA = 1;
export const ACK = 2;
export const ACK_BLOCK = 3;
export const MAX_ACK_RANGES = 16;
export const MAX_DATAGRAM_PAYLOAD = 1000;
const BLOCK_HEAD_SIZE = 27;

export function blockTsCapacity(nRanges) {
  return Math.floor((MAX_DATAGRAM_PAYLOAD - BLOCK_HEAD_SIZE - nRanges * 8 - 2) / 8);
}

export function encodeAckBlock(flow, ackId, largest, largestRecvTs, ackDelayMs, low, ranges, timestamps) {
  const buf = new Uint8Array(BLOCK_HEAD_SIZE + ranges.length * 8 + 2 + timestamps.length * 8);
  const v = new DataView(buf.buffer);
  v.setUint8(0, ACK_BLOCK); v.setUint8(1, flow); v.setUint32(2, ackId >>> 0); v.setUint32(6, largest >>> 0);
  v.setFloat64(10, largestRecvTs); v.setFloat32(18, ackDelayMs); v.setUint32(22, low >>> 0); v.setUint8(26, ranges.length);
  let off = BLOCK_HEAD_SIZE;
  for (const [first, last] of ranges) { v.setUint32(off, first); v.setUint32(off + 4, last); off += 8; }
  v.setUint16(off, timestamps.length); off += 2;
  for (const [seq, recvTs] of timestamps) { v.setUint32(off, seq); v.setFloat32(off + 4, recvTs - largestRecvTs); off += 8; }
  return buf;
}
export const FLOW_UP = 0; // browser -> server
export const FLOW_DOWN = 1; // server -> browser
export const DATA_HEADER_SIZE = 14;
export const ACK_SIZE = 22;

export function encodeData(flow, seq, sendTs, size) {
  const buf = new Uint8Array(Math.max(size, DATA_HEADER_SIZE));
  const v = new DataView(buf.buffer);
  v.setUint8(0, DATA);
  v.setUint8(1, flow);
  v.setUint32(2, seq >>> 0);
  v.setFloat64(6, sendTs);
  return buf;
}

export function encodeAck(flow, seq, echoSendTs, recvTs) {
  const buf = new Uint8Array(ACK_SIZE);
  const v = new DataView(buf.buffer);
  v.setUint8(0, ACK);
  v.setUint8(1, flow);
  v.setUint32(2, seq >>> 0);
  v.setFloat64(6, echoSendTs);
  v.setFloat64(14, recvTs);
  return buf;
}

export function decode(bytes) {
  const v = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  if (bytes.byteLength >= DATA_HEADER_SIZE && bytes[0] === DATA) {
    return { type: DATA, flow: v.getUint8(1), seq: v.getUint32(2), sendTs: v.getFloat64(6), size: bytes.byteLength };
  }
  if (bytes.byteLength >= ACK_SIZE && bytes[0] === ACK) {
    return { type: ACK, flow: v.getUint8(1), seq: v.getUint32(2), echoSendTs: v.getFloat64(6), recvTs: v.getFloat64(14) };
  }
  if (bytes.byteLength >= BLOCK_HEAD_SIZE && bytes[0] === ACK_BLOCK) {
    const largestRecvTs = v.getFloat64(10);
    const n = v.getUint8(26);
    let off = BLOCK_HEAD_SIZE;
    const ranges = [];
    for (let i = 0; i < n; i++) { ranges.push([v.getUint32(off), v.getUint32(off + 4)]); off += 8; }
    const nTs = v.getUint16(off); off += 2;
    const timestamps = [];
    for (let i = 0; i < nTs; i++) { timestamps.push([v.getUint32(off), largestRecvTs + v.getFloat32(off + 4)]); off += 8; }
    return {
      type: ACK_BLOCK, flow: v.getUint8(1), ackId: v.getUint32(2), largest: v.getUint32(6), largestRecvTs,
      ackDelayMs: v.getFloat32(18), low: v.getUint32(22), ranges, timestamps,
    };
  }
  return null;
}

File: server/protocol.py
"""Binary datagram format shared with web/protocol.js (network byte order).

DATA: type u8 | flow u8 | seq u32 | send_ts f64 (ms, sender clock) | padding
ACK:  type u8 | flow u8 | seq u32 | echo_send_ts f64 | recv_ts f64 (ms, receiver clock)
TRAIN: type u8 | flow u8 | train_id u16 | index u16 | train_len u16 | send_ts f64 | padding
  A burst of `train_len` packets sent back to back; the receiver measures the
  inter-arrival times (see server/trains.py).
PING: type u8 | flow u8 | seq u32 | send_ts f64 (sender clock) | 0 f64 | padding
PONG: type u8 | flow u8 | seq u32 | send_ts f64 (echoed) | echo_recv_ts f64 (responder clock)
  The steady-stream RTT monitor (server/rtt.py): the responder turns a PING
  around immediately, stamping its own arrival time, so one exchange gives the
  round trip *and* each leg separately (up to the clock offset).
ACK_BLOCK (one datagram acknowledging many packets):
      type u8 | flow u8 | ack_id u32 | largest u32 | largest_recv_ts f64 | ack_delay_ms f32 |
      low u32 | n_ranges u8 | n_ranges x (first u32, last u32), descending |
      n_ts u16 | n_ts x (seq u32, recv_offset_ms f32)   # recv_ts - largest_recv_ts
  Ranges list received sequence numbers in [low, largest]; anything in that
  span not covered by a range has not arrived. Timestamps cover packets newly
  received since the previous ACK_BLOCK.

`flow` names the direction of the DATA the packet refers to, so an ACK for a
server->browser packet carries FLOW_DOWN.
"""

import struct
from typing import List, NamedTuple, Optional, Tuple, Union

DATA = 1
ACK = 2
ACK_BLOCK = 3
TRAIN = 4
PING = 5
PONG = 6

FLOW_UP = 0  # browser -> server
FLOW_DOWN = 1  # server -> browser

_DATA = struct.Struct("!BBId")
_ACK = struct.Struct("!BBIdd")
_BLOCK_HEAD = struct.Struct("!BBIIdfIB")
_RANGE = struct.Struct("!II")
_U16 = struct.Struct("!H")
_TS = struct.Struct("!If")
_TRAIN = struct.Struct("!BBHHHd")
_ECHO = struct.Struct("!BBIdd")

MAX_ACK_RANGES = 16
MAX_DATAGRAM_PAYLOAD = 1000  # fits Chrome's 1024-byte maxDatagramSize with the WebTransport prefix

DATA_HEADER_SIZE = _DATA.size
TRAIN_HEADER_SIZE = _TRAIN.size
ACK_SIZE = _ACK.size
ECHO_HEADER_SIZE = _ECHO.size


class Data(NamedTuple):
    flow: int
    seq: int
    send_ts: float
    size: int


class Train(NamedTuple):
    flow: int
    train_id: int
    index: int
    train_len: int
    send_ts: float
    size: int


def encode_train(flow: int, train_id: int, index: int, train_len: int, send_ts: float, size: int) -> bytes:
    header = _TRAIN.pack(TRAIN, flow, train_id & 0xFFFF, index, train_len, send_ts)
    return header + bytes(max(0, size - TRAIN_HEADER_SIZE))


class Ping(NamedTuple):
    flow: int
    seq: int
    send_ts: float  # initiator's clock, when it handed the packet to the transport
    size: int


class Pong(NamedTuple):
    flow: int
    seq: int
    send_ts: float  # echoed unchanged, so RTT needs no state at the initiator
    echo_recv_ts: float  # responder's clock, when the PING arrived
    size: int


def encode_ping(flow: int, seq: int, send_ts: float, size: int = ECHO_HEADER_SIZE) -> bytes:
    header = _ECHO.pack(PING, flow, seq & 0xFFFFFFFF, send_ts, 0.0)
    return header + bytes(max(0, size - ECHO_HEADER_SIZE))


def encode_pong(
    flow: int, seq: int, send_ts: float, echo_recv_ts: float, size: int = ECHO_HEADER_SIZE
) -> bytes:
    header = _ECHO.pack(PONG, flow, seq & 0xFFFFFFFF, send_ts, echo_recv_ts)
    return header + bytes(max(0, size - ECHO_HEADER_SIZE))


class Ack(NamedTuple):
    flow: int
    seq: int
    echo_send_ts: float
    recv_ts: float


class AckBlock(NamedTuple):
    flow: int
    ack_id: int
    largest: int
    largest_recv_ts: float
    ack_delay_ms: float
    low: int
    ranges: List[Tuple[int, int]]  # (first, last), descending
    timestamps: List[Tuple[int, float]]  # (seq, recv_ts)


def block_ts_capacity(n_ranges: int) -> int:
    return (MAX_DATAGRAM_PAYLOAD - _BLOCK_HEAD.size - n_ranges * _RANGE.size - _U16.size) // _TS.size


def encode_ack_block(
    flow: int,
    ack_id: int,
    largest: int,
    largest_recv_ts: float,
    ack_delay_ms: float,
    low: int,
    ranges: List[Tuple[int, int]],
    timestamps: List[Tuple[int, float]],
) -> bytes:
    parts = [_BLOCK_HEAD.pack(ACK_BLOCK, flow, ack_id, largest, largest_recv_ts, ack_delay_ms, low, len(ranges))]
    parts += [_RANGE.pack(first, last) for first, last in ranges]
    parts.append(_U16.pack(len(timestamps)))
    parts += [_TS.pack(seq, recv_ts - largest_recv_ts) for seq, recv_ts in timestamps]
    return b"".join(parts)


def encode_data(flow: int, seq: int, send_ts: float, size: int) -> bytes:
    header = _DATA.pack(DATA, flow, seq & 0xFFFFFFFF, send_ts)
    return header + bytes(max(0, size - DATA_HEADER_SIZE))


def encode_ack(flow: int, seq: int, echo_send_ts: float, recv_ts: float) -> bytes:
    return _ACK.pack(ACK, flow, seq, echo_send_ts, recv_ts)


def decode(buf: bytes) -> Optional[Union[Data, Train, Ack, AckBlock]]:
    if not buf:
        return None
    if buf[0] == DATA and len(buf) >= DATA_HEADER_SIZE:
        _, flow, seq, ts = _DATA.unpack_from(buf)
        return Data(flow, seq, ts, len(buf))
    if buf[0] == TRAIN and len(buf) >= TRAIN_HEADER_SIZE:
        _, flow, train_id, index, train_len, ts = _TRAIN.unpack_from(buf)
        return Train(flow, train_id, index, train_len, ts, len(buf))
    if buf[0] in (PING, PONG) and len(buf) >= ECHO_HEADER_SIZE:
        kind, flow, seq, send_ts, echo_ts = _ECHO.unpack_from(buf)
        if kind == PING:
            return Ping(flow, seq, send_ts, len(buf))
        return Pong(flow, seq, send_ts, echo_ts, len(buf))
    if buf[0] == ACK and len(buf) >= ACK_SIZE:
        _, flow, seq, echo, recv = _ACK.unpack_from(buf)
        return Ack(flow, seq, echo, recv)
    if buf[0] == ACK_BLOCK and len(buf) >= _BLOCK_HEAD.size:
        _, flow, ack_id, largest, largest_ts, delay, low, n_ranges = _BLOCK_HEAD.unpack_from(buf)
        off = _BLOCK_HEAD.size
        ranges = []
        for _ in range(n_ranges):
            ranges.append(_RANGE.unpack_from(buf, off))
            off += _RANGE.size
        (n_ts,) = _U16.unpack_from(buf, off)
        off += _U16.size
        timestamps = []
        for _ in range(n_ts):
            seq, offset = _TS.unpack_from(buf, off)
            timestamps.append((seq, largest_ts + offset))
            off += _TS.size
        return AckBlock(flow, ack_id, largest, largest_ts, delay, low, ranges, timestamps)
    return None
